File: API/client/src/app/shared/models/order.ts
import { Address } from "./user";

export interface OrderToCreate {
    basketId: string;
    deliveryMethodId: string;
    shipToAddress: Address;
}

export interface OrderItem {
    productId: string;
    productName: string;
    pictureUrl: string;
    price: number;
    quantity: number;
}

export interface Order {
    id: string;
    buyerEmail: string;
    orderDate: Date;
    shipToAddress: Address;
    deliveryMethod: string;
    shippingPrice: number;
    orderItems: OrderItem[];
    subtotal: number;
    total: number;
    status: string;
}

File: Core/Entities/OrderAggregate/Order.cs
namespace Core.Entities.OrderAggregate;

public class Order : BaseEntity
{
    public string BuyerEmail { get; set; }
    public DateTime OrderData { get; set; } = DateTime.UtcNow;
    public IReadOnlyList<OrderItem> OrderItems { get; set; }
    public Address ShipToAddress { get; set; }
    public DeliveryMethod DeliveryMethod { get; set; }
    public decimal Subtotal { get; set; }
    public OrderStatus Status { get; set; } = OrderStatus.Pending;
    public string PaymentIntentId { get; set; }

    public Order()
    {
    }

    public Order(IReadOnlyList<OrderItem> orderItems, string buyerEmail, Address shipToAddress,
        DeliveryMethod deliveryMethod, decimal subtotal, string paymentIntentId)
    {
        BuyerEmail = buyerEmail;
        OrderItems = orderItems;
        ShipToAddress = shipToAddress;
        DeliveryMethod = deliveryMethod;
        Subtotal = subtotal;
        PaymentIntentId = paymentIntentId;
    }


    public decimal GetTotal()
    {
        return Subtotal + DeliveryMethod.Price;
    }
}
